How-To: k8s pod patching (#800)
radius-project/docs: docs/content/guides/author-apps/kubernetes/how-to-patch-pod/index.md

---
type: docs
title: "How-To: Patch Kubernetes resources using PodSpec"
linkTitle: "Patch using PodSpec"
description: "Learn how to patch Kubernetes resources using PodSpec definitions"
weight: 300
slug: 'patch-podspec'
categories: "How-To"
tags: ["containers","Kubernetes"]
---

This how-to guide will provide an overview of how to:

- Patch Radius-created Kubernetes resources using [PodSpec](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#PodSpec) definitions

## Prerequisites

- [rad CLI]({{< ref getting-started >}})
- [Radius initialized with `rad init`]({{< ref howto-environment >}})
- [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)

> ## Step 1: Define an app and a container
> 
> Begin by creating a file named `app.bicep` with a Radius application and [container]({{< ref "guides/author-apps/containers" >}}):

Reviewer: ```suggestion
## Step 1: Define a container

Begin by creating a file named `app.bicep` with a Radius [container]({{< ref "guides/author-apps/containers" >}}):
```

{{< rad file="snippets/patch-container.bicep" embed=true >}}

## Step 2: Deploy the app and container

1. Deploy and run your app:

   ```bash

>    rad run ./app.bicep

Reviewer: Without an explicit "Run rad init in a new directory and specify 'y' we'll need to pass in the application at run:
```suggestion
   rad run ./app.bicep -a demo
```

   ```

   Once the deployment completes successfully, you should see the following confirmation message along with some system logs:

   ```
   Deployment Complete

   Resources:
      demo            Applications.Core/containers

   Starting log stream...

   + demo-df76d886c-sngm8 › demo
   demo-df76d886c-sngm8 demo Using in-memory store: no connection string found
   demo-df76d886c-sngm8 demo Server is running at http://localhost:3000
   demo-df76d886c-sngm8 demo [port-forward] connected from localhost:3000 -> ::3000
   ```

   Access the application by opening [http://localhost:3000](http://localhost:3000) in a browser, where you should see the demo app:

   <img src="./demoapp-howtopatchpod.png" alt="Screenshot of Radius Demo app" width="600"/>

   <br> When you're ready to move on to the next step, use `CTRL` + `C` to exit the command.

> 2. Run the following command, making sure to specify the Kubernetes namespace to which you deployed your app:

Reviewer 2: How will the user figure this out? I'd suggest adding an `-A` to the command and then they won't need to know.

Author: I thought about this, but was afraid they'd have a bunch of namespaces and `-a` would output too large of a list. But I suppose that's unlikely given that one of the prereqs is to set up a cluster to play around with, so they might start with something that's fairly fresh. 

**tl;dr is that I'll add it :)**

> 2. Run the following command, making sure to specify the Kubernetes namespace to which you deployed your app:

Reviewer: Can we also add a quick explanation about what the `-o` flag does. Maybe "Run the following command to get the running pods and their name, status, container names, and image names."

User without much kubectl knowledge may see a bunch of commands and think they're important to read and not just a display optimmization.

Author: I assumed that most would have some knowledge of kubectl (or at least k8s) given the use case, but I agree `-o` is obscure enough to warrant more description

   ```bash
   kubectl get pods -n <YOUR_KUBERNETES_NAMESPACE> -l app.kubernetes.io/name=demo -o custom-columns=POD:.metadata.name,STATUS:.status.phase,CONTAINER_NAMES:spec.containers[:].name,CONTAINER_IMAGES:spec.containers[:].image
   ```

   You should see output confirming that a single container named `demo` was deployed and is running in your pod, similar to the following:

   ```
   POD                    STATUS    CONTAINER_NAMES   CONTAINER_IMAGES

>    demo-df76d886c-9p4gv   Running   demo              radius.azurecr.io/tutorial/webapp:edge

Reviewer: Pls use the new GHCR image

```suggestion
   demo-df76d886c-9p4gv   Running   demo              ghcr.io/radius-project/samples/demo:latest
```

   ```

## Step 3: Add a PodSpec to the container definition

> Add the following [`runtimes`]({{< ref "reference/resource-schema/core-schema/container-schema#runtimes" >}}) under `properties` of the container definition in your `app.bicep` file:

Reviewer: I'd reword this slightly to talk about what this section is a little more. Don't want this to seem like a strongly typed property in Radius, but a way to access any part of the podspec and punch through the abstraction.

```suggestion
Add the following [`runtimes`]({{< ref "reference/resource-schema/core-schema/container-schema#runtimes" >}}) configuration to the container definition in your `app.bicep` file. This allows you to punch through the Radius abstraction and directly apply any part of the Kubernetes PodSpec. In this example you're adding an additional sidecar container:
```

Author: good suggestion, edited slightly

{{< rad file="snippets/patch-runtime.bicep" embed=true markdownConfig="{linenos=table,hl_lines=[\"16-28\"]}" >}}

> Remember to save your `app.bicep` file after you've made the above changes.

## Step 4: Redeploy your app with the PodSpec added

1. Deploy and run your app again:

   ```bash

>    rad run ./app.bicep

Reviewer: ```suggestion
   rad run ./app.bicep -a demo
```

   demo-547d7dc77f-nmqpk log-collector
   ```

   Access the application by opening [http://localhost:3000](http://localhost:3000) in a browser, where you should see the demo app again unchanged from before:

   <img src="./demoapp-howtopatchpod.png" alt="Screenshot of Radius Demo app" width="600"/>

   <br> When you're ready to move on to the next step, use `CTRL` + `C` to exit the command.

2. Run the following command, making sure to specify the Kubernetes namespace to which you deployed your app:

   ```bash
   kubectl get pods -n <YOUR_KUBERNETES_NAMESPACE> -l app.kubernetes.io/name=demo -o custom-columns=POD:.metadata.name,STATUS:.status.phase,CONTAINER_NAMES:spec.containers[:].name,CONTAINER_IMAGES:spec.containers[:].image
   ```

   You should now see in the output the original `demo` app container as before, but also an additional `log-collector` container that is running in your pod, similar to the following:

   ```
   POD                     STATUS    CONTAINER_NAMES      CONTAINER_IMAGES
   demo-547d7dc77f-nmqpk   Running   log-collector,demo   ghcr.io/radius-project/fluent-bit:2.1.8,radius.azurecr.io/tutorial/webapp:edge

>    ```

Reviewer 2: This is a really cool way of demonstrating the effect. 

Depending on the speed the user does this, they might see old pods with a state of `Terminating`. You might want to mention that, and say that it's normal.

Author: good suggestion, done

   The `log-collector` container was deployed using the PodSpec definition you added to your `app.bicep` file in the `runtimes` property you added, and is now running alongside your original `demo` app container.

## Cleanup

Run the following command to [delete]({{< ref "guides/deploy-apps/howto-delete" >}}) your app and container:
   
   ```bash

>    rad app delete

Reviewer: ```suggestion
   rad app delete demo
```

---
type: docs
title: "How-To: Patch Kubernetes resources using PodSpec"
linkTitle: "Patch with PodSpec"
description: "Learn how to patch Kubernetes resources using PodSpec definitions"
weight: 400
slug: 'patch-podspec'
categories: "How-To"
tags: ["containers","Kubernetes"]
---

This how-to guide will provide an overview of how to:

> - Mount an ephemeral (short-lived) volume to a container

Reviewer 2: What makes it short-lived?

Author: PR still in draft, this was copied from another how to that I just haven't removed yet :)

## Prerequisites

- [rad CLI]({{< ref getting-started >}})
- [Radius environment]({{< ref "/guides/deploy-apps/environments/overview" >}})

## Step 1: Define an app and a container

Begin by creating a file named `app.bicep` with a Radius application and [container]({{< ref "guides/author-apps/containers" >}}):

{{< rad file="snippets/1-app.bicep" embed=true >}}

The `quickstarts/volumes` container will display the status and contents of the `/tmpdir` directory within the container.

## Step 2: Deploy the app and container

1. Deploy your app with the command:

   ```bash
   rad deploy ./app.bicep
   ```
1. Once complete, port forward to your container with [`rad resource expose`]({{< ref rad_resource_expose >}}):

   ```bash
   rad resource expose containers mycontainer -a myapp --port 5000

>    ```

Reviewer 2: Can we use `rad run` instead?

Author: Yes, I plan to do that - this was copied from another how-to, I haven't pushed my latest yet. Still in draft, please do not review again until I mark it as ready!
